### PR TITLE
Set the HSI levels in a central place

### PR DESCRIPTION
**Pull request by hughsie**
This means we don't need to worry about changing multiple
implementations if the HSI levels change for a specific ID.

It also means we can fake HSI results in the future without having
to also store the 'correct' level in the input file.

Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [ ] Feature
- [ ] Documentation

### CODE REVIEW

**Review by superm1 — DISMISSED** [review]
Exactly what I had in mind.  I didn't double check you got the mappings right (will assume you did :))

**Comment by hughsie** [comment]
Well, I "single checked" them. 🙂 I'll do some comparisons of hsi JSON output on a few machines on Monday.

File: libfwupdplugin/fu-security-attrs.c [hunk 1]
```diff
@@ -313,6 +313,72 @@ fu_security_attrs_sort_cb(gconstpointer item1, gconstpointer item2)
...
 
+static struct {
+	const gchar *appstream_id;
+	FwupdSecurityAttrLevel level;
+} appstream_id_level_map[] = {
+    {FWUPD_SECURITY_ATTR_ID_AMD_ROLLBACK_PROTECTION, FWUPD_SECURITY_ATTR_LEVEL_CRITICAL},
+    {FWUPD_SECURITY_ATTR_ID_AMD_SPI_REPLAY_PROTECTION, FWUPD_SECURITY_ATTR_LEVEL_THEORETICAL},
+    {FWUPD_SECURITY_ATTR_ID_AMD_SPI_WRITE_PROTECTION, FWUPD_SECURITY_ATTR_LEVEL_IMPORTANT},
+    {FWUPD_SECURITY_ATTR_ID_ENCRYPTED_RAM, FWUPD_SECURITY_ATTR_LEVEL_SYSTEM_PROTECTION},
+    {FWUPD_SECURITY_ATTR_ID_ENCRYPTED_RAM, FWUPD_SECURITY_ATTR_LEVEL_SYSTEM_PROTECTION},
```
Comment (superm1) [inline libfwupdplugin/fu-security-attrs.c:324]: this one is duplicated

Comment (hughsie) [reply libfwupdplugin/fu-security-attrs.c:324]: Fixed.

File: libfwupdplugin/fu-security-attrs.c [hunk 2]
```diff
@@ -313,6 +313,72 @@ fu_security_attrs_sort_cb(gconstpointer item1, gconstpointer item2)
...
+    {FWUPD_SECURITY_ATTR_ID_IOMMU, FWUPD_SECURITY_ATTR_LEVEL_IMPORTANT},
+    {FWUPD_SECURITY_ATTR_ID_MEI_MANUFACTURING_MODE, FWUPD_SECURITY_ATTR_LEVEL_CRITICAL},
+    {FWUPD_SECURITY_ATTR_ID_MEI_OVERRIDE_STRAP, FWUPD_SECURITY_ATTR_LEVEL_CRITICAL},
+    {FWUPD_SECURITY_ATTR_ID_MEI_VERSION, FWUPD_SECURITY_ATTR_LEVEL_CRITICAL},
+    {FWUPD_SECURITY_ATTR_ID_PLATFORM_DEBUG_ENABLED, FWUPD_SECURITY_ATTR_LEVEL_CRITICAL},
+    {FWUPD_SECURITY_ATTR_ID_PLATFORM_DEBUG_LOCKED, FWUPD_SECURITY_ATTR_LEVEL_IMPORTANT},
+    {FWUPD_SECURITY_ATTR_ID_PLATFORM_DEBUG_LOCKED, FWUPD_SECURITY_ATTR_LEVEL_IMPORTANT},
+    {FWUPD_SECURITY_ATTR_ID_PLATFORM_FUSED, FWUPD_SECURITY_ATTR_LEVEL_CRITICAL},
+    {FWUPD_SECURITY_ATTR_ID_PREBOOT_DMA_PROTECTION, FWUPD_SECURITY_ATTR_LEVEL_THEORETICAL},
+    {FWUPD_SECURITY_ATTR_ID_PREBOOT_DMA_PROTECTION, FWUPD_SECURITY_ATTR_LEVEL_THEORETICAL},
```
[the remark this answers is not in the document]
Comment (hughsie) [reply libfwupdplugin/fu-security-attrs.c:342]: Fixed.

File: libfwupdplugin/fu-security-attrs.c [hunk 3]
```diff
@@ -313,6 +313,72 @@ fu_security_attrs_sort_cb(gconstpointer item1, gconstpointer item2)
...
+    {FWUPD_SECURITY_ATTR_ID_INTEL_CET_ACTIVE, FWUPD_SECURITY_ATTR_LEVEL_THEORETICAL},
+    {FWUPD_SECURITY_ATTR_ID_INTEL_CET_ENABLED, FWUPD_SECURITY_ATTR_LEVEL_THEORETICAL},
+    {FWUPD_SECURITY_ATTR_ID_INTEL_SMAP, FWUPD_SECURITY_ATTR_LEVEL_SYSTEM_PROTECTION},
+    {FWUPD_SECURITY_ATTR_ID_IOMMU, FWUPD_SECURITY_ATTR_LEVEL_IMPORTANT},
+    {FWUPD_SECURITY_ATTR_ID_MEI_MANUFACTURING_MODE, FWUPD_SECURITY_ATTR_LEVEL_CRITICAL},
+    {FWUPD_SECURITY_ATTR_ID_MEI_OVERRIDE_STRAP, FWUPD_SECURITY_ATTR_LEVEL_CRITICAL},
+    {FWUPD_SECURITY_ATTR_ID_MEI_VERSION, FWUPD_SECURITY_ATTR_LEVEL_CRITICAL},
+    {FWUPD_SECURITY_ATTR_ID_PLATFORM_DEBUG_ENABLED, FWUPD_SECURITY_ATTR_LEVEL_CRITICAL},
+    {FWUPD_SECURITY_ATTR_ID_PLATFORM_DEBUG_LOCKED, FWUPD_SECURITY_ATTR_LEVEL_IMPORTANT},
+    {FWUPD_SECURITY_ATTR_ID_PLATFORM_DEBUG_LOCKED, FWUPD_SECURITY_ATTR_LEVEL_IMPORTANT},
```
[the remark this answers is not in the document]
Comment (hughsie) [reply libfwupdplugin/fu-security-attrs.c:339]: Fixed.

**Comment by hughsie** [comment]
@superm1 interestingly there is a difference; I think we actually fixed a bug too:

	--- old.json	2022-07-15 19:43:48.466585076 +0100
	+++ new.json	2022-07-15 19:48:55.649183171 +0100
	@@ -38,17 +38,6 @@ Host Security ID: HSI:1 (v1.8.3)
	       ]
	     },
	     {
	-      "AppstreamId" : "org.fwupd.hsi.Uefi.SecureBoot",
	-      "HsiLevel" : 0,
	-      "HsiResult" : "enabled",
	-      "Name" : "UEFI secure boot",
	-      "Plugin" : "uefi_capsule",
	-      "Uri" : "https://fwupd.github.io/libfwupdplugin/hsi.html#org.fwupd.hsi.Uefi.SecureBoot",
	-      "Flags" : [
	-        "success"
	-      ]
	-    },
	-    {
	       "AppstreamId" : "org.fwupd.hsi.Fwupd.Plugins",
	       "HsiLevel" : 0,
	       "HsiResult" : "not-tainted",
	@@ -216,6 +205,17 @@ Host Security ID: HSI:1 (v1.8.3)
	       ]
	     },
	     {
	+      "AppstreamId" : "org.fwupd.hsi.Uefi.SecureBoot",
	+      "HsiLevel" : 1,
	+      "HsiResult" : "enabled",
	+      "Name" : "UEFI secure boot",
	+      "Plugin" : "uefi_capsule",
	+      "Uri" : "https://fwupd.github.io/libfwupdplugin/hsi.html#org.fwupd.hsi.Uefi.SecureBoot",
	+      "Flags" : [
	+        "success"
	+      ]
	+    },
	+    {
	       "AppstreamId" : "org.fwupd.hsi.IntelBootguard.Enabled",
	       "HsiLevel" : 2,
	       "HsiResult" : "enabled",

See https://fwupd.github.io/libfwupdplugin/hsi.html#uefi-secureboot -- I think this PR is correct.

**Review by superm1 — APPROVED** [review]

**Comment by hughsie** [comment]
@superm1 so this actually broke gnome-control center, and I'm wondering if we did the right thing here.

We've used HSI level 0 for things that make the calculation invalid, e.g. kernel is invalid or there's an out-of-tree fwupd plugin, and used HSI-1 for actual platform protections. You could argue Secure Boot could be either of those. I'm erring on saying SB=false should be a HSI-1 failure as you could argue the same about any of the other protections e.g. BootGuard being turned off. The difference here is that SB is a config option in the BIOS.

I think we should leave `org.fwupd.hsi.Uefi.SecureBoot` being HSI-1, and say that `"HsiLevel" : 0` should be reserved for things that do not have any mapping to firmware security, e.g. `org.fwupd.hsi.Kernel.Swap` and all the others that have the flag `runtime-issue`.

**Comment by superm1** [comment]
> @superm1 so this actually broke gnome-control center, and I'm wondering if we did the right thing here.

Broke gnome-control-center?  Is there a gitlab link to the breakage?

> I think we should leave org.fwupd.hsi.Uefi.SecureBoot being HSI-1, and say that "HsiLevel" : 0 should be reserved for things that do not have any mapping to firmware security, e.g. org.fwupd.hsi.Kernel.Swap and all the others that have the flag runtime-issue.

Runtime issues give a `!` by design.  I think that we shouldn't have anything set to level 0.  IMO If you don't get all the stuff from level 1 you get level 0.

**Comment by hughsie** [comment]
> Broke gnome-control-center? Is there a gitlab link to the breakage?

https://gitlab.gnome.org/GNOME/gnome-control-center/-/merge_requests/1397

> Runtime issues give a ! by design

Right, thanks for the sanity check.

**Pull request merged**